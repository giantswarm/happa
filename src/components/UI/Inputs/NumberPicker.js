import PropTypes from 'prop-types';
import React from 'react';
import RUMActionTarget from 'RUM/RUMActionTarget';
import { RUMActions } from 'shared/constants/realUserMonitoring';
import { css } from 'styled-components';
import styled from 'styled-components';
import ValidationErrorMessage from 'UI/Inputs/ValidationErrorMessage';
import { mergeActionNames } from 'utils/realUserMonitoringUtils';

import TextInput from './TextInput';

// NumberPicker is a component that allows a user to pick a number by
// incrementing / decrementing a value or typing it straight into the input
// field.
//
// It only allows whole numbers.
//
// State explanation:
//
// inputValue - The current value of the input field. This might not be a valid
//              number since users should be able to clear the input before they
//              type in a new value.
//
// value - The derived value of the number picker. This will always be a valid
//         number and will be set to 0 if the the input value is invalid.
//
// valid - A boolean that represents whether or not the inputvalue is valid.
//
// validationError - A string that explains why the inputValue is not valid.
//
//
// Why do we need both inputValue and value?
// Because when a user wants to type a number, he has to first clear the
// current value. Say a user wants to type 30, but the input currently has the
// value 1 in it. Then for a moment the input has an empty string in it,
// which would break various things (like increment / decrement) if we were to
// use the inputs value as the true value for this number picker at all times.
//
// Lifecycle:
//
// This component contains an normal html input field. When this input field
// changes, its value is validated. Afterwards the parent component is
// notified of the change through the passed in onChange callback function.
// The onChange function recieves an object with the value and validity of
// the number picker:
//
// {
//   value: 13,
//   valid: true
// }

const Label = styled.div``;

const StyledTextInput = styled(TextInput)`
  text-align: center;
  width: 100%;
  height: 100%;
  outline: none;
  appearance: textfield;
  --moz-appearance: textfield;

  :focus {
    outline: none;
  }

  ::-webkit-inner-spin-button,
  ::-webkit-outer-spin-button {
    display: none;
    --webkit-appearance: none;
    margin: 0;
  }
`;

const Wrapper = styled.div`
  display: inline-block;
  margin-top: 4px;
  margin-bottom: 10px;
`;

const Control = styled.div`
  width: 160px;
  text-align: center;
  border-radius: 4px;
  position: relative;
`;

const IncrementDecrementButtonCSS = css`
  position: absolute;
  top: 1px;
  width: 35px;
  height: calc(100% - 2px);
  background-color: #3b5f7b;
  user-select: none;
  z-index: 1;
  display: flex;
  justify-content: center;
  align-items: center;

  &:hover {
    background-color: #486d8a;
  }

  &:active {
    background-color: #3b5f7b;
    color: #aaa;
  }

  &.disabled,
  &.disabled:hover,
  &.disabled:active {
    background-color: #567;
    color: #eee;
  }
`;

const IncrementButton = styled.div`
  ${IncrementDecrementButtonCSS};
  right: 1px;
  border-top-right-radius: 4px;
  border-bottom-right-radius: 4px;
`;

const DecrementButton = styled.div`
  ${IncrementDecrementButtonCSS};
  left: 1px;
  border-top-left-radius: 4px;
  border-bottom-left-radius: 4px;
`;

class NumberPicker extends React.Component {
  state = {
    inputValue: this.props.value,
    value: this.props.value,
    valid: true,
    validationError: '',
  };

  componentDidUpdate(prevProps) {
    if (prevProps.max !== this.props.max || prevProps.min !== this.props.min) {
      const { value, validationError } = this.validateInput(this.state.value);
      const isValid = validationError.length < 1;
      this.updateValue(value, isValid, validationError);
    }
  }

  increment = () => {
    const currentValue = this.props.value;
    const desiredValue = currentValue + this.props.stepSize;

    this.updateInput(desiredValue);
  };

  decrement = () => {
    const currentValue = this.props.value;
    const desiredValue = currentValue - this.props.stepSize;

    this.updateInput(desiredValue);
  };

  updateInput = (desiredValue, allowInvalidValues = false) => {
    // Validate.
    // eslint-disable-next-line prefer-const
    let { value, validationError } = this.validateInput(desiredValue);

    // Ensure values are never above max or below min. They can be null.
    const { max, min } = this.props;
    switch (true) {
      case value === null:
        value = '';
        break;
      case !allowInvalidValues && value < min:
        value = min;
        break;
      case !allowInvalidValues && value > max:
        value = max;
        break;
    }

    const isValid = validationError.length < 1 || (min <= 0 && value === 0);

    this.updateValue(value, isValid, validationError);
  };

  updateValue(newValue, isValid = false, validationError = '') {
    this.setState(
      {
        inputValue: newValue,
        value: newValue,
        valid: isValid,
        validationError,
      },
      () => {
        // Notify Parent.
        if (this.props.onChange) {
          this.props.onChange({
            value: this.state.value,
            valid: this.state.valid,
          });
        }
      }
    );
  }

  validateInput = (desiredValue) => {
    if (desiredValue === '') {
      return {
        value: null,
        validationError: 'Field must not be empty',
      };
    } else if (desiredValue > this.props.max) {
      return {
        value: parseInt(desiredValue),
        validationError: `Value must not be larger than ${this.props.max}`,
      };
    } else if (desiredValue < this.props.min) {
      return {
        value: parseInt(desiredValue),
        validationError: `Value must not be smaller than ${this.props.min}`,
      };
    } else if (!isWholeNumber(parseFloat(desiredValue))) {
      return {
        value: parseInt(desiredValue),
        validationError: 'Value must be a whole number',
      };
    }

    return {
      value: parseInt(desiredValue),
      validationError: '',
    };
  };

  handleFocus = (event) => {
    event.target.select();
  };

  render() {
    return (
      <Wrapper
        className={`${this.props.className ?? ''} ${this.props.theme ?? ''}`}
      >
        {this.props.label ? <Label>{this.props.label}</Label> : undefined}

        <Control>
          {this.props.readOnly ? undefined : (
            <RUMActionTarget
              name={mergeActionNames(
                RUMActions.DecrementNumber,
                this.props.eventNameSuffix
              )}
            >
              <DecrementButton
                className={
                  this.state.inputValue === this.props.min && 'disabled'
                }
                onClick={this.decrement}
                aria-label='Decrement'
                role='button'
              >
                &ndash;
              </DecrementButton>
            </RUMActionTarget>
          )}
          <StyledTextInput
            disabled={this.props.readOnly}
            onChange={(e) => this.updateInput(e.target.value, true)}
            onFocus={this.handleFocus}
            step={this.props.stepSize}
            type='number'
            value={
              this.props.readOnly ? this.props.value : this.state.inputValue
            }
            title={this.props.title}
          />
          {this.props.readOnly ? undefined : (
            <RUMActionTarget
              name={mergeActionNames(
                RUMActions.IncrementNumber,
                this.props.eventNameSuffix
              )}
            >
              <IncrementButton
                className={this.props.value === this.props.max && 'disabled'}
                onClick={this.increment}
                aria-label='Increment'
                role='button'
              >
                +
              </IncrementButton>
            </RUMActionTarget>
          )}
        </Control>
        <ValidationErrorMessage message={this.state.validationError} />
      </Wrapper>
    );
  }
}

NumberPicker.propTypes = {
  label: PropTypes.string,
  value: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
  stepSize: PropTypes.number,
  min: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
  max: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
  onChange: PropTypes.func,
  readOnly: PropTypes.bool,
  theme: PropTypes.string,
  className: PropTypes.string,
  title: PropTypes.string,
  /** This string is appended to the user action event names recorded in Real User Monitoring. Should be UPPERCASE. */
  eventNameSuffix: PropTypes.string,
};

function isWholeNumber(value) {
  if (typeof value === 'number' && value % 1 === 0) {
    return true;
  }

  return false;
}

export default NumberPicker;
