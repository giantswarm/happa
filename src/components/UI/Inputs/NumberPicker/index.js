import usePrevious from 'lib/hooks/usePrevious';
import PropTypes from 'prop-types';
import React, { useCallback, useEffect, useState } from 'react';
import RUMActionTarget from 'RUM/RUMActionTarget';
import { RUMActions } from 'shared/constants/realUserMonitoring';
import { css } from 'styled-components';
import styled from 'styled-components';
import ValidationErrorMessage from 'UI/Inputs/ValidationErrorMessage';
import { mergeActionNames } from 'utils/realUserMonitoringUtils';

import TextInput from '../TextInput';

function validateInput(desiredValue, min, max) {
  switch (true) {
    case desiredValue === '':
      return 'Field must not be empty';
    case max && desiredValue > max:
      return `Value must not be larger than ${max}`;
    case min && desiredValue < min:
      return `Value must not be smaller than ${min}`;
    case !isWholeNumber(desiredValue):
      return 'Value must be a whole number';
    default:
      return '';
  }
}

function isWholeNumber(value) {
  return value % 1 === 0;
}

const Label = styled.div``;

const StyledTextInput = styled(TextInput)`
  text-align: center;
  width: 100%;
  height: 100%;
  outline: none;
  appearance: textfield;
  --moz-appearance: textfield;

  :focus {
    outline: none;
  }

  ::-webkit-inner-spin-button,
  ::-webkit-outer-spin-button {
    display: none;
    --webkit-appearance: none;
    margin: 0;
  }
`;

const Wrapper = styled.div`
  display: inline-block;
  margin-top: 4px;
  margin-bottom: 10px;
`;

const Control = styled.div`
  width: 160px;
  text-align: center;
  border-radius: 4px;
  position: relative;
`;

const IncrementDecrementButtonCSS = css`
  position: absolute;
  top: 1px;
  width: 35px;
  height: calc(100% - 2px);
  background-color: #3b5f7b;
  user-select: none;
  z-index: 1;
  display: flex;
  justify-content: center;
  align-items: center;

  &:hover {
    background-color: #486d8a;
  }

  &:active {
    background-color: #3b5f7b;
    color: #aaa;
  }

  &.disabled,
  &.disabled:hover,
  &.disabled:active {
    background-color: #567;
    color: #eee;
  }
`;

const IncrementButton = styled.div`
  ${IncrementDecrementButtonCSS};
  right: 1px;
  border-top-right-radius: 4px;
  border-bottom-right-radius: 4px;
`;

const DecrementButton = styled.div`
  ${IncrementDecrementButtonCSS};
  left: 1px;
  border-top-left-radius: 4px;
  border-bottom-left-radius: 4px;
`;

/**
 * A component that allows a user to pick a number by
 * incrementing / decrementing a value or typing it
 * straight into the input field.
 */
const NumberPicker = ({
  min,
  max,
  label,
  className,
  theme,
  readOnly,
  eventNameSuffix,
  onChange,
  value,
  stepSize,
  title,
}) => {
  const [currValue, setCurrValue] = useState(value);
  const [validationError, setValidationError] = useState('');

  const inputValue = Number.isNaN(currValue) ? '' : currValue;

  const prevMin = usePrevious(min);
  const prevMax = usePrevious(max);

  const updateValue = useCallback(
    (newValue, error = '') => {
      setCurrValue(newValue);
      setValidationError(error);

      const isValid = error.length < 1;

      onChange?.({
        value: newValue,
        valid: isValid,
      });
    },
    [onChange]
  );

  const updateInput = (desiredValue, allowInvalidValues = false) => {
    const error = validateInput(desiredValue, min, max);
    let newValue = desiredValue;

    switch (true) {
      case !allowInvalidValues && newValue < min:
        newValue = min;
        break;
      case !allowInvalidValues && newValue > max:
        newValue = max;
        break;
    }

    updateValue(newValue, error);
  };

  const increment = () => {
    const desiredValue = currValue + stepSize;

    updateInput(desiredValue);
  };

  const decrement = () => {
    const desiredValue = currValue - stepSize;

    updateInput(desiredValue);
  };

  const handleFocus = (event) => {
    event.target.select();
  };

  useEffect(() => {
    if (prevMax !== max || prevMin !== min) {
      const error = validateInput(currValue, min, max);
      updateValue(currValue, error);
    }
  }, [min, max, prevMin, prevMax, currValue, updateValue]);

  return (
    <Wrapper className={`${className ?? ''} ${theme ?? ''}`}>
      {label && <Label>{label}</Label>}

      <Control>
        {!readOnly && (
          <RUMActionTarget
            name={mergeActionNames(RUMActions.DecrementNumber, eventNameSuffix)}
          >
            <DecrementButton
              className={currValue === min && 'disabled'}
              onClick={decrement}
              aria-label='Decrement'
              role='button'
            >
              &ndash;
            </DecrementButton>
          </RUMActionTarget>
        )}

        <StyledTextInput
          disabled={readOnly}
          onChange={(e) => updateInput(e.target.valueAsNumber, true)}
          onFocus={handleFocus}
          step={stepSize}
          type='number'
          value={inputValue}
          title={title}
        />

        {!readOnly && (
          <RUMActionTarget
            name={mergeActionNames(RUMActions.IncrementNumber, eventNameSuffix)}
          >
            <IncrementButton
              className={currValue === max && 'disabled'}
              onClick={increment}
              aria-label='Increment'
              role='button'
            >
              +
            </IncrementButton>
          </RUMActionTarget>
        )}
      </Control>
      <ValidationErrorMessage message={validationError} />
    </Wrapper>
  );
};

NumberPicker.propTypes = {
  label: PropTypes.string,
  value: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
  stepSize: PropTypes.number,
  min: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
  max: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
  onChange: PropTypes.func,
  readOnly: PropTypes.bool,
  theme: PropTypes.string,
  className: PropTypes.string,
  title: PropTypes.string,
  /** This string is appended to the user action event names recorded in Real User Monitoring. Should be UPPERCASE. */
  eventNameSuffix: PropTypes.string,
};

export default NumberPicker;
