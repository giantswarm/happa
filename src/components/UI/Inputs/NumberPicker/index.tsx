import usePrevious from 'lib/hooks/usePrevious';
import PropTypes from 'prop-types';
import React, { useCallback, useEffect, useState } from 'react';
import { css } from 'styled-components';
import styled from 'styled-components';

import TextInput from '../TextInput';

function validateInput(desiredValue: number, min?: number, max?: number) {
  switch (true) {
    case Number.isNaN(desiredValue):
      return 'Field must not be empty';
    case typeof max !== 'undefined' && desiredValue > max:
      return `Value must not be larger than ${max}`;
    case typeof min !== 'undefined' && desiredValue < min:
      return `Value must not be smaller than ${min}`;
    case !isWholeNumber(desiredValue):
      return 'Value must be a whole number';
    default:
      return '';
  }
}

function isWholeNumber(value: number) {
  return value % 1 === 0;
}

const StyledTextInput = styled(TextInput)`
  text-align: center;
  outline: none;
  appearance: textfield;
  --moz-appearance: textfield;

  :focus {
    outline: none;
  }

  ::-webkit-inner-spin-button,
  ::-webkit-outer-spin-button {
    display: none;
    --webkit-appearance: none;
    margin: 0;
  }
`;

const Controls = styled.div`
  position: absolute;
  top: 1px;
  left: 0;
  width: 100%;
  height: calc(100% - 2px);
  display: flex;
  justify-content: space-between;
  pointer-events: none;
  z-index: 1;
`;

const IncrementDecrementButtonCSS = css`
  position: relative;
  width: 35px;
  height: 100%;
  background-color: ${({ theme }) => theme.global.colors.border.dark};
  user-select: none;

  display: flex;
  justify-content: center;
  align-items: center;
  pointer-events: all;

  &:hover {
    background-color: ${({ theme }) =>
      theme.global.colors['background-front'].dark};
  }

  &:active {
    opacity: 0.4;
  }

  &.disabled,
  &.disabled:hover,
  &.disabled:active {
    color: ${({ theme }) => theme.global.colors.text.dark};
    cursor: default;
    opacity: 0.8;

    ::after {
      content: '';
      position: absolute;
      left: 0;
      top: 0;
      width: 100%;
      height: 100%;
      background-color: ${({ theme }) =>
        theme.global.colors['status-disabled']};
      opacity: 0.3;
    }
  }
`;

const IncrementButton = styled.div`
  ${IncrementDecrementButtonCSS};
  right: 1px;
  border-top-right-radius: 4px;
  border-bottom-right-radius: 4px;
`;

const DecrementButton = styled.div`
  ${IncrementDecrementButtonCSS};
  left: 1px;
  border-top-left-radius: 4px;
  border-bottom-left-radius: 4px;
`;

interface INumberPickerProps
  extends Omit<React.ComponentPropsWithoutRef<typeof TextInput>, 'onChange'> {
  value?: number;
  min?: number;
  max?: number;
  step?: number;
  onChange?: (patch: { value: number; valid: boolean }) => void;
}

/**
 * A component that allows a user to pick a number by
 * incrementing / decrementing a value or typing it
 * straight into the input field.
 */
const NumberPicker = React.forwardRef<HTMLInputElement, INumberPickerProps>(
  ({ value, min, max, step, readOnly, onChange, ...props }, ref) => {
    const [currValue, setCurrValue] = useState<number>(value!);
    const [validationError, setValidationError] = useState('');

    const inputValue = Number.isNaN(currValue) ? '' : currValue;

    const prevMin = usePrevious(min);
    const prevMax = usePrevious(max);

    const updateValue = useCallback(
      (newValue: number, error = '') => {
        setCurrValue(newValue);
        setValidationError(error);

        const isValid = error.length < 1;

        onChange?.({
          value: newValue,
          valid: isValid,
        });
      },
      [onChange]
    );

    const updateInput = (desiredValue: number, allowInvalidValues = false) => {
      const error = validateInput(desiredValue, min, max);
      let newValue = desiredValue;

      switch (true) {
        case !allowInvalidValues &&
          typeof min !== 'undefined' &&
          newValue < min:
          newValue = min!;
          break;
        case !allowInvalidValues &&
          typeof max !== 'undefined' &&
          newValue > max:
          newValue = max!;
          break;
      }

      updateValue(newValue, error);
    };

    const increment = () => {
      const desiredValue = currValue + step!;

      updateInput(desiredValue);
    };

    const decrement = () => {
      const desiredValue = currValue - step!;

      updateInput(desiredValue);
    };

    const handleFocus = (event: React.FocusEvent<HTMLInputElement>) => {
      event.target.select();
    };

    useEffect(() => {
      if (prevMax !== max || prevMin !== min) {
        const error = validateInput(currValue, min, max);
        updateValue(currValue, error);
      }
    }, [min, max, prevMin, prevMax, currValue, updateValue]);

    return (
      <StyledTextInput
        readOnly={readOnly}
        onChange={(e) => updateInput(e.target.valueAsNumber, true)}
        onFocus={handleFocus}
        step={step}
        type='number'
        value={inputValue}
        error={validationError}
        {...props}
        ref={ref}
      >
        {!readOnly && (
          <Controls>
            <DecrementButton
              className={currValue === min ? 'disabled' : undefined}
              onClick={decrement}
              aria-label='Decrement'
              role='button'
            >
              &ndash;
            </DecrementButton>
            <IncrementButton
              className={currValue === max ? 'disabled' : undefined}
              onClick={increment}
              aria-label='Increment'
              role='button'
            >
              +
            </IncrementButton>
          </Controls>
        )}
      </StyledTextInput>
    );
  }
);

NumberPicker.propTypes = {
  label: PropTypes.string,
  value: PropTypes.number,
  step: PropTypes.number,
  min: PropTypes.number,
  max: PropTypes.number,
  onChange: PropTypes.func,
  readOnly: PropTypes.bool,
  className: PropTypes.string,
  title: PropTypes.string,
};

NumberPicker.defaultProps = {
  value: 0,
  step: 1,
  min: 0,
  max: 999,
};

export default NumberPicker;
