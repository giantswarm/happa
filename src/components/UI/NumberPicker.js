import { css } from '@emotion/core';
import styled from '@emotion/styled';
import PropTypes from 'prop-types';
import React from 'react';
import ValidationErrorMessage from 'UI/ValidationErrorMessage';

// NumberPicker is a component that allows a user to pick a number by
// incrementing / decrementing a value or typing it straight into the input
// field.
//
// It only allows whole numbers.
//
// State explanation:
//
// inputValue - The current value of the input field. This might not be a valid
//              number since users should be able to clear the input before they
//              type in a new value.
//
// value - The derived value of the number picker. This will always be a valid
//         number and will be set to 0 if the the input value is invalid.
//
// valid - A boolean that represents whether or not the inputvalue is valid.
//
// validationError - A string that explains why the inputValue is not valid.
//
//
// Why do we need both inputValue and value?
// Because when a user wants to type a number, he has to first clear the
// current value. Say a user wants to type 30, but the input currently has the
// value 1 in it. Then for a moment the input has an empty string in it,
// which would break various things (like increment / decrement) if we were to
// use the inputs value as the true value for this number picker at all times.
//
// Lifecycle:
//
// This component contains an normal html input field. When this input field
// changes, its value is validated. Afterwards the parent component is
// notified of the change through the passed in onChange callback function.
// The onChange function recieves an object with the value and validity of
// the number picker:
//
// {
//   value: 13,
//   valid: true
// }

const Label = styled.div`
  display: inline-block;
  width: 150px;
`;

const ValueSpan = styled.span`
  input {
    display: inline-block;
    background-color: inherit;
    border: none;
    text-align: center;
    width: 100%;
    height: 100%;
    outline: none;
    font-size: 16px;
  }
  input:focus {
    outline: none;
  }
  input[type='number']::-webkit-inner-spin-button,
  input[type='number']::-webkit-outer-spin-button {
    --webkit-appearance: none;
    margin: 0;
  }
`;

const Wrapper = styled.div`
  display: inline-block;
  margin-top: 4px;
  margin-bottom: 10px;

  /* to hide the increment/decrement buttons when disabled */
  input:disabled {
    appearance: textfield;
  }
`;

const Control = styled.div`
  width: 160px;
  display: inline-block;
  height: 35px;
  background-color: #32526a;
  line-height: 35px;
  text-align: center;
  border-radius: 5px;
  position: relative;
  margin-bottom: 5px;
`;

const IncrementDecrementButtonCSS = css`
  position: absolute;
  display: inline-block;
  top: 0px;
  width: 35px;
  background-color: #3b5f7b;
  cursor: pointer;
  user-select: none;
  &:hover {
    background-color: #486d8a;
  }
  &:active {
    background-color: #3b5f7b;
    color: #aaa;
  }
  &.disabled,
  &.disabled:hover,
  &.disabled:active {
    background-color: #567;
    color: #eee;
  }
`;

const IncrementButton = styled.div`
  ${IncrementDecrementButtonCSS};
  right: 0px;
  border-top-right-radius: 5px;
  border-bottom-right-radius: 5px;
`;

const DecrementButton = styled.div`
  ${IncrementDecrementButtonCSS};
  left: 0px;
  border-top-left-radius: 5px;
  border-bottom-left-radius: 5px;
`;

class NumberPicker extends React.Component {
  state = {
    inputValue: this.props.value,
    value: this.props.value,
    valid: true,
    validationError: '',
  };

  increment = () => {
    const currentValue = this.props.value;
    const desiredValue = currentValue + this.props.stepSize;

    this.updateInput({
      target: { value: desiredValue },
    });
  };

  decrement = () => {
    const currentValue = this.props.value;
    const desiredValue = currentValue - this.props.stepSize;

    this.updateInput({
      target: { value: desiredValue },
    });
  };

  updateInput = (e) => {
    const desiredValue = e.target.value;

    // Validate.
    // eslint-disable-next-line prefer-const
    let { value, validationError } = this.validateInput(desiredValue);

    // Ensure values are never above max or below min. They can be null.
    const { max, min } = this.props;
    value = value === null ? '' : value < min ? min : value > max ? max : value;

    // Update state.
    this.setState(
      {
        inputValue: value,
        value,
        valid: Boolean(value) || (min <= 0 && value === 0),
        validationError,
      },
      () => {
        // Notify Parent.
        if (this.props.onChange) {
          this.props.onChange({
            value: this.state.value,
            valid: this.state.valid,
          });
        }
      }
    );
  };

  validateInput = (desiredValue) => {
    if (desiredValue === '') {
      return {
        value: null,
        validationError: 'Field must not be empty',
      };
    } else if (desiredValue > this.props.max) {
      return {
        value: parseInt(desiredValue),
        validationError: `Value must not be larger than ${this.props.max}`,
      };
    } else if (desiredValue < this.props.min) {
      return {
        value: parseInt(desiredValue),
        validationError: `Value must not be smaller than ${this.props.min}`,
      };
    } else if (!isWholeNumber(parseFloat(desiredValue))) {
      return {
        value: parseInt(desiredValue),
        validationError: 'Value must be a whole number',
      };
    }

    return {
      value: parseInt(desiredValue),
      validationError: '',
    };
  };

  handleFocus = (event) => {
    event.target.select();
  };

  render() {
    return (
      <Wrapper
        className={`${this.props.className ?? ''} ${this.props.theme ?? ''} ${
          this.props.readOnly ? 'readonly ' : ''
        }`}
      >
        {this.props.label ? <Label>{this.props.label}</Label> : undefined}

        <Control>
          {this.props.readOnly ? undefined : (
            <DecrementButton
              className={this.state.inputValue === this.props.min && 'disabled'}
              onClick={this.decrement}
            >
              &ndash;
            </DecrementButton>
          )}
          <ValueSpan>
            <input
              disabled={this.props.readOnly}
              // max={this.props.max}
              // min={this.props.min}
              onChange={this.updateInput}
              onFocus={this.handleFocus}
              step={this.props.stepSize}
              type='number'
              value={
                this.props.readOnly ? this.props.value : this.state.inputValue
              }
            />
          </ValueSpan>
          {this.props.readOnly ? undefined : (
            <IncrementButton
              className={this.props.value === this.props.max && 'disabled'}
              onClick={this.increment}
            >
              +
            </IncrementButton>
          )}
        </Control>
        <ValidationErrorMessage message={this.state.validationError} />
      </Wrapper>
    );
  }
}

NumberPicker.propTypes = {
  label: PropTypes.string,
  value: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
  stepSize: PropTypes.number,
  min: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
  max: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
  onChange: PropTypes.func,
  readOnly: PropTypes.bool,
  theme: PropTypes.string,
  className: PropTypes.string,
};

function isWholeNumber(value) {
  if (typeof value === 'number' && value % 1 === 0) {
    return true;
  }

  return false;
}

export default NumberPicker;
