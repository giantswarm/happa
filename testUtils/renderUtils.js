import { render } from '@testing-library/react';
import App from 'App';
import { ConnectedRouter } from 'connected-react-router';
import { ThemeProvider } from 'emotion-theming';
import { createMemoryHistory } from 'history';
import React from 'react';
import { Provider } from 'react-redux';
import { AppRoutes } from 'shared/constants/routes';
import configureStore from 'stores/configureStore';
import theme from 'styles/theme';

export const initialStorage = {
  user:
    '{"email":"developer@giantswarm.io","auth":{"scheme":"giantswarm","token":"a-valid-token"},"isAdmin":true}',
};

/**
 * This function will render the whole app with a mocked store in the route
 * provided.
 * @param {String} route.
 * @param {HTMLElement} container An HTMLElement.
 */
export function renderRouteWithStore(
  initialRoute = AppRoutes.Home,
  state = {},
  storage = initialStorage
) {
  localStorage.replaceWith(storage);

  const history = createMemoryHistory({
    initialEntries: [initialRoute],
    initialIndex: 0,
  });

  const store = configureStore(state, history);

  const app = render(<App {...{ store, theme, history }} />);

  return app;
}

/**
 * Render component in the Theme context
 * @param {React.ReactType} component The React Component
 * @param {<P extends Record<string, any>>React.ComponentProps<P>} props Props to pass to the component
 * @param {<Q extends TestingLibrary.Queries>TestingLibrary.RenderOptions<Q>} [options] Testing library render options
 */
export function renderWithTheme(component, props, options) {
  return render(getComponentWithTheme(component, props), options);
}

/**
 * Include component in the Theme context and return it
 * @param {React.ReactType} Component The React Component
 * @param {<P extends Record<string, any>>React.ComponentProps<P>} props Props to pass to the component
 */
export function getComponentWithTheme(Component, props) {
  return (
    <ThemeProvider theme={theme}>
      <Component {...props} />
    </ThemeProvider>
  );
}

/**
 * Render component in the Store context
 * @param {React.ReactType} component The React Component
 * @param {<P extends Record<string, any>>React.ComponentProps<P>} props Props to pass to the component
 * @param {Record<string, any>} state Current Store state
 * @param {<Q extends TestingLibrary.Queries>TestingLibrary.RenderOptions<Q>} [options] Testing library render options
 */
export function renderWithStore(component, props, state, options) {
  return render(getComponentWithStore(component, props, state), options);
}

/**
 * Include component in the Store context and return it
 * @param {React.ReactType} Component The React Component
 * @param {<P extends Record<string, any>>React.ComponentProps<P>} props Props to pass to the component
 * @param {Record<string, any>} state Current Store state
 * @param {Record<string, any>} storage Current LocalStorage value
 * @param {History} history Current Browser history
 */
export function getComponentWithStore(
  Component,
  props = {},
  state = {},
  storage = initialStorage,
  history = createMemoryHistory()
) {
  localStorage.replaceWith(storage);

  const store = configureStore(state, history);

  const app = (
    <Provider store={store}>
      <ThemeProvider theme={theme}>
        <ConnectedRouter history={history}>
          <Component {...props} />
        </ConnectedRouter>
      </ThemeProvider>
    </Provider>
  );

  return app;
}
