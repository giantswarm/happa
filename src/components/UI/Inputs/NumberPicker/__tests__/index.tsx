import { renderWithTheme } from 'test/renderUtils';

import NumberPicker from '..';

describe('NumberPicker', () => {
  it('renders a simple input', () => {
    const { container } = renderWithTheme(NumberPicker, {
      value: 3,
      id: 'money',
      label: 'Count your money',
      info: 'Aha! Did not know that',
      help: 'A very helpful message',
    });
    expect(container.firstChild).toMatchSnapshot();
  });
});
