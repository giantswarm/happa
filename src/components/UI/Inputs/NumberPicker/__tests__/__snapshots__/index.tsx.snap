// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NumberPicker renders a simple input 1`] = `
<div
  class="StyledGrommet-sc-19lkkz7-0 giQQcM"
>
  <div
    class="ThemeProvider__AppWrapper-sc-12xd2cf-0 kuNpSX"
  >
    <div
      class="StyledBox-sc-13pk1d4-0 gPuWHK FormField__FormFieldBox-sc-m9hood-0 kVQSBU"
    >
      <label
        class="StyledText-sc-1sadyjn-0 eqjVSW"
        for="money"
      >
        Count your money
      </label>
      <span
        class="StyledText-sc-1sadyjn-0 fqoiau"
      >
        A very helpful message
      </span>
      <div
        class="StyledBox-sc-13pk1d4-0 kbQwEP FormField__FormFieldContentBox-sc-m9hood-1"
      >
        <div
          class="StyledTextInput__StyledTextInputContainer-sc-1x30a0s-1 jqwtzG"
        >
          <input
            autocomplete="off"
            class="StyledTextInput-sc-1x30a0s-0 kmYDQh NumberPicker__StyledTextInput-sc-h062ap-0 cfcBUA"
            id="money"
            step="1"
            type="number"
            value="3"
          />
        </div>
        <div
          class="NumberPicker__Controls-sc-h062ap-1 dmihKA"
        >
          <div
            aria-label="Decrement"
            class="NumberPicker__IncrementDecrementButton-sc-h062ap-2 kcrxCA"
            role="button"
          >
            â€“
          </div>
          <div
            aria-label="Increment"
            class="NumberPicker__IncrementDecrementButton-sc-h062ap-2 kcrxCA"
            role="button"
          >
            +
          </div>
        </div>
      </div>
      <span
        class="StyledText-sc-1sadyjn-0 heWbGB"
      >
        Aha! Did not know that
      </span>
    </div>
  </div>
</div>
`;
